src/smtencoding/FStar.SMTEncoding.Solver.fs

```diff
@@ -106,43 +106,44 @@ let with_hints_db fname f =
 let filter_using_facts_from (e:env) (theory:decls_t) =
     let should_enc_fid fid =
         match fid with
-        | Namespace lid -> Env.should_enc_lid e lid
-        | _ -> false
+        | Name lid -> Env.should_enc_lid e lid
+        | _ -> failwith "Solver.fs::filter_using_facts_from expected only Name fact ids"
     in
     let matches_fact_ids (include_assumption_names:BU.smap<bool>) (a:Term.assumption) =
       match a.assumption_fact_ids with
       | [] -> true //retaining `a` because it is not tagged with a fact id
       | _ ->
-        a.assumption_fact_ids |> BU.for_some should_enc_fid
+        a.assumption_fact_ids |> List.filter (function | Name _ -> true | _ -> false) |> BU.for_some should_enc_fid
```

Comment:
Why the double scan to filter first to Names and then to check should_enc_fid? Can't this be done in a single pass?